### ISSUE TITLE
On OSX cannot upload as "needs a json library"

### ISSUE THREAD

**Issue by justincormack** [keplerproject/luarocks#559]
I installed on OSX with `brew install lua` but I cannot upload as it says `Error: A JSON library is required for this command.` Any idea which json library it needs? The `brew` package should install what it needs.

**Comment by hishamhm** [comment]
This happens on all platforms... Yeah, I know it's not the best UX. LuaRocks needs a simple JSON library able to `json.encode` and `json.decode`, and we didn't want to pre-install any particular one... Just run something like `luarocks install dkjson` or `luarocks install lua-cjson` and you'll be set.

**Comment by mpeterv** [comment]
(FTR the message is improved on maser branch)

**Comment by justincormack** [comment]
Can you at least add this to the install instructions then?

**Comment by hishamhm** [comment]
@mpeterv and it's been in the master branch for a while... perhaps it's about time for a new release? Is anything currently holding us back? ( @Tieske , @ignacio , @siffiejoe )

**Comment by mpeterv** [comment]
@hishamhm I don't think there is anything incomplete to hold us back. I've got a few PRs but they can wait. Although I'd love to get at least the download progress thingy (#554) included in the release =D

**Comment by ignacio** [comment]
I'm sorry I won't be of any help for a little while.

**Comment by daurnimator** [comment]
> @mpeterv and it's been in the master branch for a while... perhaps it's about time for a new release? Is anything currently holding us back?

I'd like to see #465 and #511 addressed.

**Comment by Tieske** [comment]
> This happens on all platforms... Yeah, I know it's not the best UX. LuaRocks needs a simple JSON library able to json.encode and json.decode, and we didn't want to pre-install any particular one... Just run something like luarocks install dkjson or luarocks install lua-cjson and you'll be set.

Can't we just include a simple pure-lua one in the LuaRocks installation? eg. pe-parser is included (albeit only on Windows during installation)

**Comment by hishamhm** [comment]
> Can't we just include a simple pure-lua one in the LuaRocks installation?

We could, but it gets me torn between "LR should be self-contained" and "shouldn't we start eating our own dogfood and fetch the stuff we need as rocks?" Currently, we're doing a little bit of each, but copping out and telling the user to luarocks-install luasec and a JSON library (without picking a horse in the race-of-JSON-libraries).

I wonder what people's perceptions would be if LR installed a few rocks from the get-go during its own installation. If we open this floodgate, even LR itself could be made more modular. But would people complain and perceive this as bloat?

**Comment by daurnimator** [comment]
> Currently, we're doing a little bit of each, but copping out and telling the user to luarocks-install luasec and a JSON library (without picking a horse in the race-of-JSON-libraries).
> 
> I wonder what people's perceptions would be if LR installed a few rocks from the get-go during its own installation. If we open this floodgate, even LR itself could be made more modular. But would people complain and perceive this as bloat?

I would be okay with this with one limitation: "bundled rocks" should have *no* external C library dependencies. If possible, prefer libraries that are pure-lua (or at least work without dependencies).

We already depend on lfs; so that would be immediate candidates for inclusion.
For json libraries, we should prefer dkjson over cjson (as dkjson has a pure-lua variant).
It would disqualify luasec (as it requires openssl to be installed).

This will allow us to split off many things in the 'tools' directory: e.g. the zip-packing library (luarocks/src/tools/zip.lua) to their own projects.

**Comment by Tieske** [comment]
One concern I have with this is the separation of LuaRocks and the system itself. I wouldn't want the json library picked by luarocks to show up in my system tree. So if it pulls in stuff, it should do so in its own separate tree.

imo LR should be self contained, but may "unlock" extra features if local packages have been installed.

**Comment by daurnimator** [comment]
> I wouldn't want the json library picked by luarocks to show up in my system tree

Why not?

**Comment by Tieske** [comment]
Because there are too many and not all compatible. The exact reason we’re having this discussion...

Hence my proposal to include a simple pure lua one that does the job with the LR sources

**Comment by daurnimator** [comment]
> Because there are too many and not all compatible

But they are not mutally exclusive, and you can have more than once installed without issue. So again: why not have luarocks ship with eg the dkjson rock and have it show up in the system tree?

**Comment by siffiejoe** [comment]
Once issue to consider: the json library that comes pre-installed with LuaRocks will probably become the de-facto standard json library in the Luaverse (unless it's a private module) much like IE on Windows ....

**Comment by hishamhm** [comment]
Yes, @siffiejoe, I had this in the back of my mind when I mentioned "picking a horse in the race-of-JSON-libraries". I'm not sure this is a problem per se, though. But it's definitely something to consider.

The concerns and criteria @daurnimator brings up are also interesting. LuaSec however is the one thing current LuaRocks complains about the most, and if we don't want to depend on wget/curl forever, that would be the most obvious dependency to auto-fetch. @daurnimator, I know you have a battle going against luasocket/luasec and I appreciate any efforts in modernizing the ecosystem, but as long as cqueues is Unix-only I don't think we should adopt it in LR. I _do_ want to eventually enable `fs_use_modules` on Windows; I'd rather include a set of rocks in the package than the bunch of UnxUtils executables.

**Comment by daurnimator** [comment]
>  LuaSec however is the one thing current LuaRocks complains about the most,

I don't think it's reasonable to bundle luasec; or infact anything that depends on OpenSSL.
OpenSSL cannot be assumed for many platforms; and infact, apple seem to be doing their best to make it hard to use.

> if we don't want to depend on wget/curl forever, 

What's wrong with that? I guess the maintenance burden?
Infact on windows we should be able to rely on `bitsadmin` in windows 7 and above.

Though I *don't* see an issue with splitting out the 'wget downloader' to a separate lua module; and then bundling that with luarocks.

> as long as cqueues is Unix-only I don't think we should adopt it in LR

I somewhat reluctantly agree. I am working on this; but I don't think I'll see the end of it within the next year.

> I do want to eventually enable fs_use_modules on Windows; I'd rather include a set of rocks in the package than the bunch of UnxUtils executables.

I'm crossing my fingers that midipix can be a solution here..... more to come in the next months.

**Comment by mpeterv** [comment]
@hishamhm do you plan a new release? Or should we wait for Windows testsuite port?

**Comment by hishamhm** [comment]
@mpeterv Yes, I was about to make one but I slipped into other things. No need to wait for the Windows testsuite, I think (It's more important to get it into CI than into a release). But I'm afraid I won't be able to do it this week. I'll schedule a release for at most next week.

**Comment by mpeterv** [comment]
I think that as a solution for this luarocks should bundle a pure lua json library (so, dkjson). "upload" working out of the box is too important. For all other optional dependencies there are fallbacks.

**Comment by redcatbear** [comment]
While I am usually a big fan of dependency-separation, I have to agree in this particular case that bundling would be an option. Alternatively it would help if the combination of the command `upload` and the switch `local` worked. That would at least allow using `upload` without root privileges.
